## 1. Importing Libraries

In [126]:
import os
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split
import warnings
warnings.filterwarnings("ignore")

## 2.Reading the data

In [127]:
Project_Dir = r'C:\Users\admin\OneDrive\Desktop\Flights_SageMaker_Project'

Data_Dir = 'Data'

In [128]:
os.path.join(Project_Dir, Data_Dir, 'Data_Train.csv')

'C:\\Users\\admin\\OneDrive\\Desktop\\Flights_SageMaker_Project\\Data\\Data_Train.csv'

In [129]:
def get_data(name):
    file_name = f"{name}.csv"
    file_path = os.path.join(Project_Dir,Data_Dir,file_name)
    return pd.read_csv(file_path)

In [130]:
flights = get_data('Data_Train')
flights

Airline Date_of_Journey    Source Destination  \
0           IndiGo      24-03-2019  Banglore   New Delhi   
1        Air India      01-05-2019   Kolkata    Banglore   
2      Jet Airways      09-06-2019     Delhi      Cochin   
3           IndiGo      12-05-2019   Kolkata    Banglore   
4           IndiGo      01-03-2019  Banglore   New Delhi   
...            ...             ...       ...         ...   
10678     Air Asia      09-04-2019   Kolkata    Banglore   
10679    Air India      27-04-2019   Kolkata    Banglore   
10680  Jet Airways      27-04-2019  Banglore       Delhi   
10681      Vistara      01-03-2019  Banglore   New Delhi   
10682    Air India      09-05-2019     Delhi      Cochin   

                       Route Dep_Time      Arrival_Time Duration Total_Stops  \
0                  BLR ? DEL    22:20  22-03-2024 01:10   2h 50m    non-stop   
1      CCU ? IXR ? BBI ? BLR    05:50             13:15   7h 25m     2 stops   
2      DEL ? LKO ? BOM ? COK    09:25  10-06-2024 04:25      19h     2 stops   
3            CCU ? NAG ? BLR    18:05             23:30   5h 25m      1 stop   
4            BLR ? NAG ? DEL    16:50             21:35   4h 45m      1 stop   
...                      ...      ...               ...      ...         ...   
10678              CCU ? BLR    19:55             22:25   2h 30m    non-stop   
10679              CCU ? BLR    20:45             23:20   2h 35m    non-stop   
10680              BLR ? DEL    08:20             11:20       3h    non-stop   
10681              BLR ? DEL    11:30             14:10   2h 40m    non-stop   
10682  DEL ? GOI ? BOM ? COK    10:55             19:15   8h 20m     2 stops   

      Additional_Info  Price  
0             No info   3897  
1             No info   7662  
2             No info  13882  
3             No info   6218  
4             No info  13302  
...               ...    ...  
10678         No info   4107  
10679         No info   4145  
10680         No info   7229  
10681         No info  12648  
10682         No info  11753  

[10683 rows x 11 columns]

In [131]:
flights.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10683 entries, 0 to 10682
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   Airline          10683 non-null  object
 1   Date_of_Journey  10683 non-null  object
 2   Source           10683 non-null  object
 3   Destination      10683 non-null  object
 4   Route            10682 non-null  object
 5   Dep_Time         10683 non-null  object
 6   Arrival_Time     10683 non-null  object
 7   Duration         10683 non-null  object
 8   Total_Stops      10682 non-null  object
 9   Additional_Info  10683 non-null  object
 10  Price            10683 non-null  int64 
dtypes: int64(1), object(10)
memory usage: 918.2+ KB


- The dataset containes 10,683 rows and 11 features.
- Columns 'Rout' and 'Total_Stops' has missing values.
- There are some features wont be treated as datetime data type.

In [132]:
flights.isnull().sum()

Airline            0
Date_of_Journey    0
Source             0
Destination        0
Route              1
Dep_Time           0
Arrival_Time       0
Duration           0
Total_Stops        1
Additional_Info    0
Price              0
dtype: int64

## 3. Preliminary Analysis

### 3.1 Check Data Types

In [133]:
flights.head()

Airline Date_of_Journey    Source Destination                  Route  \
0       IndiGo      24-03-2019  Banglore   New Delhi              BLR ? DEL   
1    Air India      01-05-2019   Kolkata    Banglore  CCU ? IXR ? BBI ? BLR   
2  Jet Airways      09-06-2019     Delhi      Cochin  DEL ? LKO ? BOM ? COK   
3       IndiGo      12-05-2019   Kolkata    Banglore        CCU ? NAG ? BLR   
4       IndiGo      01-03-2019  Banglore   New Delhi        BLR ? NAG ? DEL   

  Dep_Time      Arrival_Time Duration Total_Stops Additional_Info  Price  
0    22:20  22-03-2024 01:10   2h 50m    non-stop         No info   3897  
1    05:50             13:15   7h 25m     2 stops         No info   7662  
2    09:25  10-06-2024 04:25      19h     2 stops         No info  13882  
3    18:05             23:30   5h 25m      1 stop         No info   6218  
4    16:50             21:35   4h 45m      1 stop         No info  13302

In [134]:
flights.dtypes

Airline            object
Date_of_Journey    object
Source             object
Destination        object
Route              object
Dep_Time           object
Arrival_Time       object
Duration           object
Total_Stops        object
Additional_Info    object
Price               int64
dtype: object

In [135]:
value = flights.Date_of_Journey.iloc[6]
value

'12-03-2019'

In [136]:
value = flights.Dep_Time.iloc[6]
value

'18:55'

In [137]:
value = flights.Duration.iloc[6]
value

'15h 30m'

### 3.2 Check for Duplicates

In [138]:
flights.duplicated().sum()

220

In [139]:
# lets check the duplicate data by applying filter

flights[flights.duplicated(keep=False) == True].sort_values(['Airline','Date_of_Journey','Source','Destination'])

Airline Date_of_Journey    Source Destination                  Route  \
6321  Air India      01-03-2019  Banglore   New Delhi  BLR ? BOM ? AMD ? DEL   
9848  Air India      01-03-2019  Banglore   New Delhi  BLR ? BOM ? AMD ? DEL   
1495  Air India      01-04-2019   Kolkata    Banglore  CCU ? DEL ? COK ? BLR   
9913  Air India      01-04-2019   Kolkata    Banglore  CCU ? DEL ? COK ? BLR   
3598  Air India      01-05-2019   Kolkata    Banglore  CCU ? GAU ? DEL ? BLR   
...         ...             ...       ...         ...                    ...   
2692   SpiceJet      24-03-2019  Banglore   New Delhi              BLR ? DEL   
2870   SpiceJet      24-03-2019  Banglore   New Delhi              BLR ? DEL   
3711   SpiceJet      24-03-2019  Banglore   New Delhi              BLR ? DEL   
2634    Vistara      24-03-2019  Banglore   New Delhi              BLR ? DEL   
5470    Vistara      24-03-2019  Banglore   New Delhi              BLR ? DEL   

     Dep_Time      Arrival_Time Duration Total_Stops  \
6321    08:50  02-03-2024 23:55   39h 5m     2 stops   
9848    08:50  02-03-2024 23:55   39h 5m     2 stops   
1495    10:00  02-04-2024 01:20  15h 20m     2 stops   
9913    10:00  02-04-2024 01:20  15h 20m     2 stops   
3598    09:50  02-05-2024 08:55   23h 5m     2 stops   
...       ...               ...      ...         ...   
2692    05:45             08:35   2h 50m    non-stop   
2870    05:45             08:35   2h 50m    non-stop   
3711    20:30             23:20   2h 50m    non-stop   
2634    11:30             14:10   2h 40m    non-stop   
5470    11:30             14:10   2h 40m    non-stop   

                   Additional_Info  Price  
6321                       No info  17135  
9848                       No info  17135  
1495                       No info  10408  
9913                       No info  10408  
3598                       No info  13227  
...                            ...    ...  
2692  No check-in baggage included   4273  
2870  No check-in baggage included   4273  
3711  No check-in baggage included   3873  
2634                       No info   5403  
5470                       No info   5403  

[415 rows x 11 columns]

### 3.3 Observations

>- The datatype of feature 'Date_of_Journey' should be datetime.
>- Same we get stuff at feature 'Dep_Time','Arrival_time'.
>- The type of 'duration' is mixed so we have to handle them.
>- 'Total_Stops' is also mixed and we can convert them to int or float.
>- There are 220 duplicatse this should be removed.

## 4. Detailed Analysis

### Airline

In [140]:
flights.Airline

0             IndiGo
1          Air India
2        Jet Airways
3             IndiGo
4             IndiGo
            ...     
10678       Air Asia
10679      Air India
10680    Jet Airways
10681        Vistara
10682      Air India
Name: Airline, Length: 10683, dtype: object

In [141]:
flights.Airline.unique()

array(['IndiGo', 'Air India', 'Jet Airways', 'SpiceJet',
       'Multiple carriers', 'GoAir', 'Vistara', 'Air Asia',
       'Vistara Premium economy', 'Jet Airways Business',
       'Multiple carriers Premium economy', 'Trujet'], dtype=object)

In [142]:
# We are putting block of code inside bracket beacuse that will makes better to read.
(
    flights
    .Airline
    .str.replace(" Premium economy","")
    .str.replace(" Business","")
    .str.title()
)

0             Indigo
1          Air India
2        Jet Airways
3             Indigo
4             Indigo
            ...     
10678       Air Asia
10679      Air India
10680    Jet Airways
10681        Vistara
10682      Air India
Name: Airline, Length: 10683, dtype: object

>- Some of the entries has inconsisent values so we need to replace those values.

### Date Of Journey

In [143]:
flights.Date_of_Journey

0        24-03-2019
1        01-05-2019
2        09-06-2019
3        12-05-2019
4        01-03-2019
            ...    
10678    09-04-2019
10679    27-04-2019
10680    27-04-2019
10681    01-03-2019
10682    09-05-2019
Name: Date_of_Journey, Length: 10683, dtype: object

In [144]:
pd.to_datetime(flights.Date_of_Journey,dayfirst=True)

0       2019-03-24
1       2019-05-01
2       2019-06-09
3       2019-05-12
4       2019-03-01
           ...    
10678   2019-04-09
10679   2019-04-27
10680   2019-04-27
10681   2019-03-01
10682   2019-05-09
Name: Date_of_Journey, Length: 10683, dtype: datetime64[ns]

### Source 

In [145]:
flights.Source.unique()

array(['Banglore', 'Kolkata', 'Delhi', 'Chennai', 'Mumbai'], dtype=object)

### Destination

In [146]:
flights.Destination.unique()

array(['New Delhi', 'Banglore', 'Cochin', 'Kolkata', 'Delhi', 'Hyderabad'],
      dtype=object)

### Departure time

In [147]:
flights.Dep_Time

0        22:20
1        05:50
2        09:25
3        18:05
4        16:50
         ...  
10678    19:55
10679    20:45
10680    08:20
10681    11:30
10682    10:55
Name: Dep_Time, Length: 10683, dtype: object

In [148]:
(
    flights
    .Dep_Time
    .loc[lambda ser: ser.str.contains("[^0-9:]")]
)

Series([], Name: Dep_Time, dtype: object)

In [149]:
pd.to_datetime(flights.Dep_Time).dt.time

0        22:20:00
1        05:50:00
2        09:25:00
3        18:05:00
4        16:50:00
           ...   
10678    19:55:00
10679    20:45:00
10680    08:20:00
10681    11:30:00
10682    10:55:00
Name: Dep_Time, Length: 10683, dtype: object

### Arrival Time

In [150]:
flights.Arrival_Time

0        22-03-2024 01:10
1                   13:15
2        10-06-2024 04:25
3                   23:30
4                   21:35
               ...       
10678               22:25
10679               23:20
10680               11:20
10681               14:10
10682               19:15
Name: Arrival_Time, Length: 10683, dtype: object

In [157]:
(
    flights
    .Arrival_Time
    .loc[lambda ser: ser.str.contains("[^0-9:]")]
    .str.split(" ",n=0)
    .str.get(1)
)

0        01:10
2        04:25
6        10:25
7        05:05
8        10:25
         ...  
10666    19:00
10667    20:20
10672    19:00
10673    04:25
10674    21:20
Name: Arrival_Time, Length: 4335, dtype: object

In [161]:
(
    flights
    .Arrival_Time
    .apply(lambda x: x.split(" ")[1] if len(x.split(" ")) > 1 else x)
)

0        01:10
1        13:15
2        04:25
3        23:30
4        21:35
         ...  
10678    22:25
10679    23:20
10680    11:20
10681    14:10
10682    19:15
Name: Arrival_Time, Length: 10683, dtype: object

### Duration

In [169]:
flights.Duration

0        2h 50m
1        7h 25m
2           19h
3        5h 25m
4        4h 45m
          ...  
10678    2h 30m
10679    2h 35m
10680        3h
10681    2h 40m
10682    8h 20m
Name: Duration, Length: 10683, dtype: object

In [171]:
(
    flights
    .Duration
    .loc[lambda ser: ~ser.str.contains("m")]
    .unique()
)

array(['19h', '23h', '22h', '12h', '3h', '5h', '10h', '18h', '24h', '15h',
       '16h', '8h', '14h', '20h', '13h', '11h', '9h', '27h', '26h', '4h',
       '7h', '30h', '21h', '28h', '47h', '6h', '25h', '38h', '34h'],
      dtype=object)

In [173]:
(
    flights
    .Duration
    .loc[lambda ser: ~ser.str.contains("h")]
)

6474    5m
Name: Duration, dtype: object

In [175]:
flights.iloc[[6474]]

Airline Date_of_Journey  Source Destination                  Route  \
6474  Air India      06-03-2019  Mumbai   Hyderabad  BOM ? GOI ? PNQ ? HYD   

     Dep_Time Arrival_Time Duration Total_Stops Additional_Info  Price  
6474    16:50        16:55       5m     2 stops         No info  17327

>- This observation indexed is clearly wrong so we can delete this row beacause as we know that there is no flight duration 5m

In [193]:
(
    flights
    .Duration
    .drop(index=6474)
    .str.split(" ", expand=True)
    .set_axis(['hour','minute'], axis=1) # where used to give an columns names
    .assign(
        hour= lambda df_:(
            df_
            .hour
            .str.replace("h","")
            .astype(int)
            .mul(60)
        ),
        minute= lambda df_:(
            df_
            .minute
            .str.replace("m","")
            .fillna("0")
            .astype(int)
        )
    )
    .sum(axis=1)
)

0         170
1         445
2        1140
3         325
4         285
         ... 
10678     150
10679     155
10680     180
10681     160
10682     500
Length: 10682, dtype: int64

In [199]:
(
    flights
    .Duration
    .drop(index=6474)
    .str.split(" ", expand=True)
    .set_axis(['hour','minute'], axis=1) # where used to give an columns names
    .assign(
        hour= lambda df_:(
            df_
            .hour
            .str.replace("h","")
            .astype(int)
            .mul(60)
        ),
        minute= lambda df_:(
            df_
            .minute
            .str.replace("m","")
            .fillna("0")
            .astype(int)
        )
    )
    .sum(axis=1)
    .rename('duration_minutes')
    .to_frame()    
    .join(flights.Duration.drop(index=6474))
)

duration_minutes Duration
0                   170   2h 50m
1                   445   7h 25m
2                  1140      19h
3                   325   5h 25m
4                   285   4h 45m
...                 ...      ...
10678               150   2h 30m
10679               155   2h 35m
10680               180       3h
10681               160   2h 40m
10682               500   8h 20m

[10682 rows x 2 columns]

### TotalStops

In [211]:
flights.Total_Stops.unique()

array(['non-stop', '2 stops', '1 stop', '3 stops', nan, '4 stops'],
      dtype=object)

In [225]:
(
    flights
    .Total_Stops
    .replace("non-stop","0")
    .str.replace(" stops?","", regex =True) # This is an regular expression that ? helps t search for pattern stop or stops so that s is an optional
    .pipe(lambda ser: pd.to_numeric(ser)) 
)

0        0.0
1        2.0
2        2.0
3        1.0
4        1.0
        ... 
10678    0.0
10679    0.0
10680    0.0
10681    0.0
10682    2.0
Name: Total_Stops, Length: 10683, dtype: float64

### Addition Info

In [236]:
flights.Additional_Info.unique()

array(['No info', 'In-flight meal not included',
       'No check-in baggage included', '1 Short layover', 'No Info',
       '1 Long layover', 'Change airports', 'Business class',
       'Red-eye flight', '2 Long layover'], dtype=object)

## 5.Cleaning Opertaions

In [205]:
def convert_to_minutes(ser):
    return(
        ser
        .str.split(" ", expand=True)
        .set_axis(['hour','minute'], axis=1) # where used to give an columns names
        .assign(
            hour= lambda df_:(
                df_
                .hour
                .str.replace("h","")
                .astype(int)
                .mul(60)
            ),
            minute= lambda df_:(
                df_
                .minute
                .str.replace("m","")
                .fillna("0")
                .astype(int)
            )
        )
        .sum(axis=1)
    )

In [237]:
def clean_data(df):
    return (
        df
        .drop(index=[6474])
        .assign(**{
            col: df[col].str.strip()
            for col in df.select_dtypes(include='O').columns
        })
        .rename(columns=str.lower)
        .assign(
            airline=lambda df_: (
                df_
                .airline
                .str.replace(" Premium economy", "")
                .str.replace(" Business", "")
                .str.title()
            ),
            date_of_journey=lambda df_: pd.to_datetime(df_.date_of_journey, dayfirst=True),
            dep_time=lambda df_: pd.to_datetime(df_.dep_time).dt.time,
            arrival_time=lambda df_: (
                df_
                .arrival_time
                .apply(lambda x: x.split(" ")[1] if len(x.split(" ")) > 1 else x)
            ),
            duration= lambda df_: df_.duration.pipe(convert_to_minutes),
            total_stops=lambda df_:(
                df_
                .total_stops
                .replace("non-stop","0")
                .str.replace(" stops?","", regex =True) # This is an regular expression that ? helps t search for pattern stop or stops so that s is an optional
                .pipe(lambda ser: pd.to_numeric(ser))
            ),
            additional_info=lambda df_: df_.additional_info.replace("No info","No Info")
        )
        .drop(columns='route')
    )

>- pipe where used in pandas to apply fun over series or dataframe

In [239]:
flights_cleaned = clean_data(flights)
flights_cleaned

airline date_of_journey    source destination  dep_time  \
0           Indigo      2019-03-24  Banglore   New Delhi  22:20:00   
1        Air India      2019-05-01   Kolkata    Banglore  05:50:00   
2      Jet Airways      2019-06-09     Delhi      Cochin  09:25:00   
3           Indigo      2019-05-12   Kolkata    Banglore  18:05:00   
4           Indigo      2019-03-01  Banglore   New Delhi  16:50:00   
...            ...             ...       ...         ...       ...   
10678     Air Asia      2019-04-09   Kolkata    Banglore  19:55:00   
10679    Air India      2019-04-27   Kolkata    Banglore  20:45:00   
10680  Jet Airways      2019-04-27  Banglore       Delhi  08:20:00   
10681      Vistara      2019-03-01  Banglore   New Delhi  11:30:00   
10682    Air India      2019-05-09     Delhi      Cochin  10:55:00   

      arrival_time  duration  total_stops additional_info  price  
0            01:10       170          0.0         No Info   3897  
1            13:15       445          2.0         No Info   7662  
2            04:25      1140          2.0         No Info  13882  
3            23:30       325          1.0         No Info   6218  
4            21:35       285          1.0         No Info  13302  
...            ...       ...          ...             ...    ...  
10678        22:25       150          0.0         No Info   4107  
10679        23:20       155          0.0         No Info   4145  
10680        11:20       180          0.0         No Info   7229  
10681        14:10       160          0.0         No Info  12648  
10682        19:15       500          2.0         No Info  11753  

[10682 rows x 10 columns]

## 6. Split the Data

In [240]:
flights_final = flights_cleaned.sample(1000)
flights_final

airline date_of_journey    source destination  dep_time  \
6706        Jet Airways      2019-05-15  Banglore       Delhi  06:00:00   
8662             Indigo      2019-05-03   Kolkata    Banglore  17:15:00   
864           Air India      2019-05-18     Delhi      Cochin  17:40:00   
3249             Indigo      2019-04-06   Kolkata    Banglore  15:15:00   
9128        Jet Airways      2019-06-12   Kolkata    Banglore  16:30:00   
...                 ...             ...       ...         ...       ...   
5014          Air India      2019-05-15   Kolkata    Banglore  10:00:00   
1326        Jet Airways      2019-05-18   Kolkata    Banglore  16:30:00   
3159             Indigo      2019-06-15  Banglore       Delhi  04:00:00   
1177        Jet Airways      2019-05-01   Kolkata    Banglore  08:25:00   
7948  Multiple Carriers      2019-03-21     Delhi      Cochin  13:15:00   

     arrival_time  duration  total_stops              additional_info  price  
6706        08:45       165          0.0                      No Info   7229  
8662        19:50       155          0.0                      No Info   4804  
864         23:00       320          1.0                      No Info   8277  
3249        17:45       150          0.0                      No Info   4174  
9128        10:05      1055          1.0  In-flight meal not included   8529  
...           ...       ...          ...                          ...    ...  
5014        13:45      1665          2.0                      No Info  16057  
1326        23:35       425          1.0                      No Info  14781  
3159        06:50       170          0.0                      No Info   4423  
1177        12:00      1655          1.0  In-flight meal not included   9663  
7948        21:20       485          2.0                      No Info   6653  

[1000 rows x 10 columns]

In [241]:
X = flights_final.drop(columns="price")
y = flights_final.price.copy()

In [242]:
X_,X_test,y_,y_test = train_test_split(X,y,test_size=0.2,random_state=42)
X_train,X_val,y_train,y_val = train_test_split(X_,y_,test_size=0.2,random_state=42)

In [245]:
print(X_train.shape, y_train.shape)
print(X_val.shape, y_val.shape)
print(X_test.shape, y_test.shape)

(640, 9) (640,)
(160, 9) (160,)
(200, 9) (200,)


## 7. Export the Subsets

In [248]:
def export_subset(X, y, name):
    file_name = f"{name}.csv"
    file_path = os.path.join(Project_Dir, Data_Dir, file_name)
    
    X.join(y).to_csv(file_path, index=False)
    
    return pd.read_csv(file_path).head()

In [249]:
export_subset(X_train, y_train, 'train')

airline date_of_journey   source destination  dep_time  \
0             Indigo      2019-04-18  Kolkata    Banglore  20:20:00   
1        Jet Airways      2019-05-21  Kolkata    Banglore  06:30:00   
2        Jet Airways      2019-05-01  Kolkata    Banglore  08:25:00   
3  Multiple Carriers      2019-06-01    Delhi      Cochin  09:15:00   
4        Jet Airways      2019-05-12  Kolkata    Banglore  08:25:00   

  arrival_time  duration  total_stops              additional_info  price  
0        22:55       155          0.0                      No Info   4174  
1        19:50       800          1.0  In-flight meal not included  10844  
2        18:15       590          1.0  In-flight meal not included  10844  
3        19:00       585          1.0                      No Info  12042  
4        08:15      1430          1.0  In-flight meal not included   8586

In [250]:
export_subset(X_test, y_test , 'test')

airline date_of_journey    source destination  dep_time arrival_time  \
0    Air India      2019-06-09     Delhi      Cochin  17:40:00        23:00   
1       Indigo      2019-05-18     Delhi      Cochin  10:35:00        21:00   
2  Jet Airways      2019-06-12     Delhi      Cochin  05:30:00        19:00   
3    Air India      2019-05-27    Mumbai   Hyderabad  06:20:00        07:40   
4       Indigo      2019-03-01  Banglore   New Delhi  13:00:00        15:50   

   duration  total_stops              additional_info  price  
0       320          1.0                      No Info   7384  
1       625          1.0                      No Info   7503  
2       810          2.0  In-flight meal not included  10368  
3        80          0.0                      No Info   3100  
4       170          0.0                      No Info  19685

In [251]:
export_subset(X_val, y_val , 'validation')

airline date_of_journey   source destination  dep_time  \
0        Jet Airways      2019-05-18  Kolkata    Banglore  20:00:00   
1        Jet Airways      2019-05-18    Delhi      Cochin  10:00:00   
2        Jet Airways      2019-03-27    Delhi      Cochin  05:10:00   
3        Jet Airways      2019-05-12  Kolkata    Banglore  21:10:00   
4  Multiple Carriers      2019-06-24    Delhi      Cochin  08:00:00   

  arrival_time  duration  total_stops              additional_info  price  
0        23:35      1655          1.0  In-flight meal not included  10844  
1        04:25      1105          1.0  In-flight meal not included  12373  
2        04:25      1395          2.0  In-flight meal not included   9412  
3        10:05       775          1.0  In-flight meal not included   8586  
4        21:00       780          1.0                      No Info  13797